File: crates/klyron_cli/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::CliBuilder::Impl {
public:
  Klyron::CliConfig config;
  bool verbose = false;
};

Klyron::CliBuilder::Klyron::CliBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::CliBuilder& Klyron::CliBuilder::withConfig(const Klyron::CliConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::CliBuilder& Klyron::CliBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::CliBuilder::Klyron::CliInstance Klyron::CliBuilder::build() {
  Klyron::CliInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_cli/bindings/rust/builder.rs
//! Builder pattern for klyron_cli

use crate::types::Klyron::CliConfig;

#[derive(Debug, Default)]
pub struct Klyron::CliBuilder {
    config: Option<Klyron::CliConfig>,
    verbose: bool,
}

impl Klyron::CliBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::CliConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::CliInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::CliInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::CliInstance {
    pub config: Klyron::CliConfig,
    pub verbose: bool,
}
